src/scheduler/MpiWorld.cpp

```diff
@@ -369,39 +367,39 @@ void MpiWorld::send(int sendRank,
     int msgId = (int)faabric::util::generateGid();
 
     // Create the message
-    faabric::MPIMessage m;
-    m.set_id(msgId);
-    m.set_worldid(id);
-    m.set_sender(sendRank);
-    m.set_destination(recvRank);
-    m.set_type(dataType->id);
-    m.set_count(count);
-    m.set_messagetype(messageType);
+    auto m = std::make_shared<faabric::MPIMessage>();
+    m->set_id(msgId);
+    m->set_worldid(id);
+    m->set_sender(sendRank);
+    m->set_destination(recvRank);
+    m->set_type(dataType->id);
+    m->set_count(count);
+    m->set_messagetype(messageType);
 
     // Work out whether the message is sent locally or to another host
     const std::string otherHost = getHostForRank(recvRank);
     bool isLocal = otherHost == thisHost;
 
     // Set up message data
     if (count > 0 && buffer != nullptr) {
-        m.set_buffer(buffer, dataType->size * count);
+        m->set_buffer(buffer, dataType->size * count);
     }
 
     // Dispatch the message locally or globally
     if (isLocal) {
         if (messageType == faabric::MPIMessage::RMA_WRITE) {
             logger->trace("MPI - local RMA write {} -> {}", sendRank, recvRank);
-            synchronizeRmaWrite(m, false);
+            synchronizeRmaWrite(*m, false);
         } else {
             logger->trace("MPI - send {} -> {}", sendRank, recvRank);
-            getLocalQueue(sendRank, recvRank)->enqueue(m);
+            getLocalQueue(sendRank, recvRank)->enqueue(std::move(m));
         }
     } else {
         logger->trace("MPI - send remote {} -> {}", sendRank, recvRank);
 
         // TODO - avoid creating a client each time?
         scheduler::FunctionCallClient client(otherHost);
-        client.sendMPIMessage(m);
+        client.sendMPIMessage(*m);
```

Comment:
Can we avoid a copy here too? If we change `sendMPIMessage` to take a `shared_ptr` then only dereference it at the last minute